osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -0,0 +1,399 @@
+                writer.Write(FormattableString.Invariant($"{tempHitSample.Volume},"));
+                writer.Write(FormattableString.Invariant($"{(timingPoint != null ? "1" : "0")},"));
+                writer.Write(FormattableString.Invariant($"{(int)effectFlags}"));
+                writer.Write("\n");
+            }
+        }
+
+        private void handleHitObjects(TextWriter writer)
+        {
+            if (beatmap.HitObjects.Count == 0)
+                return;
+
+            writer.WriteLine("[HitObjects]");
+
+            foreach (var h in beatmap.HitObjects)
+            {
+                switch (beatmap.BeatmapInfo.RulesetID)
+                {
+                    case 0:
+                        handleOsuHitObject(writer, h);
+                        break;
+
+                    case 1:
+                        handleTaikoHitObject(writer, h);
+                        break;
+
+                    case 2:
+                        handleCatchHitObject(writer, h);
+                        break;
+
+                    case 3:
+                        handleManiaHitObject(writer, h);
+                        break;
+                }
+            }
+        }
+
+        private void handleOsuHitObject(TextWriter writer, HitObject hitObject)
+        {
+            var positionData = hitObject as IHasPosition;
+            var comboData = hitObject as IHasCombo;
+
+            Debug.Assert(positionData != null);
+            Debug.Assert(comboData != null);
+
+            LegacyHitObjectType hitObjectType = (LegacyHitObjectType)(comboData.ComboOffset << 4);
+            if (comboData.NewCombo)
+                hitObjectType |= LegacyHitObjectType.NewCombo;
+            if (hitObject is IHasCurve)
+                hitObjectType |= LegacyHitObjectType.Slider;
+            else if (hitObject is IHasEndTime)
+                hitObjectType |= LegacyHitObjectType.Spinner | LegacyHitObjectType.NewCombo;
+            else
+                hitObjectType |= LegacyHitObjectType.Circle;
+
+            writer.Write(FormattableString.Invariant($"{positionData.X},"));
+            writer.Write(FormattableString.Invariant($"{positionData.Y},"));
+            writer.Write(FormattableString.Invariant($"{hitObject.StartTime},"));
+            writer.Write(FormattableString.Invariant($"{(int)hitObjectType},"));
+
+            writer.Write(hitObject is IHasCurve
+                ? FormattableString.Invariant($"0,")
+                : FormattableString.Invariant($"{(int)toLegacyHitSoundType(hitObject.Samples)},"));
+
+            if (hitObject is IHasCurve curveData)
+            {
+                PathType? lastType = null;
+
+                for (int i = 0; i < curveData.Path.ControlPoints.Count; i++)
+                {
+                    PathControlPoint point = curveData.Path.ControlPoints[i];
+
+                    if (point.Type.Value != null)
```

Comment:
I'm not sure how you feel about `continue`s, but this seems like a good place to put one and invert the check to get the path out of the way.

Reply:
I did it this way because of the `if (i != 0)` check below. Unless I'm missing something, I'll either have to dupe that block into the `continue`ing condition (which seems weird due to the code duping), or add a try-finally (which isn't saving much).